### PR TITLE
[FEATURE][callouts] Add anchor point position settings for polygon features

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This adds an anchor point position property when drawing callouts. ATM, a few anchor point positions have been added for polygon features (most interesting is the point on exterior).

image
_Left, pole of inaccessibility anchor point; right: point on exterior_

@nyalldawson , your review is needed. I'm adding tests later today but the code is ready to be reviewed.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/callouts/qgscallout.cpp [hunk 1]
```diff
@@ -38,6 +38,7 @@ void QgsCallout::initPropertyDefinitions()
     { QgsCallout::OffsetFromAnchor, QgsPropertyDefinition( "OffsetFromAnchor", QObject::tr( "Offset from feature" ), QgsPropertyDefinition::DoublePositive, origin ) },
     { QgsCallout::OffsetFromLabel, QgsPropertyDefinition( "OffsetFromLabel", QObject::tr( "Offset from label" ), QgsPropertyDefinition::DoublePositive, origin ) },
     { QgsCallout::DrawCalloutToAllParts, QgsPropertyDefinition( "DrawCalloutToAllParts", QObject::tr( "Draw lines to all feature parts" ), QgsPropertyDefinition::Boolean, origin ) },
+    { QgsCallout::AnchorPointPosition, QgsPropertyDefinition( "AnchorPointPosition", QObject::tr( "Feature's anchor point position" ), QgsPropertyDefinition::String, origin ) },
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp]: There's API here to set the permissible values too - check how it's done in labelling for an example

File: src/core/callouts/qgscallout.cpp [hunk 2]
```diff
@@ -71,6 +74,7 @@ bool QgsCallout::saveProperties( QDomDocument &doc, QDomElement &element, const
 
   QDomElement calloutElement = doc.createElement( QStringLiteral( "callout" ) );
   calloutElement.setAttribute( QStringLiteral( "type" ), type() );
+  calloutElement.setAttribute( QStringLiteral( "anchorPoint" ), encodeAnchorPoint( mAnchorPoint ) );
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp]: This shouldnt be needed - it's already covered in properties()

File: src/core/callouts/qgscallout.cpp [hunk 3]
```diff
@@ -294,7 +340,22 @@ void QgsSimpleLineCallout::draw( QgsRenderContext &context, QRectF rect, const d
...
+            line = label.shortestLine( partAnchor.poleOfInaccessibility( std::max( partAnchor.boundingBox().width(), partAnchor.boundingBox().height() ) / 20.0 ) ); // really rough (but quick) pole of inaccessibility
+            break;
+          case QgsCallout::PointOnSurface:
+            line = label.shortestLine( partAnchor.pointOnSurface() );
+            break;
+          case QgsCallout::PointOnExterior:
+            line = label.shortestLine( partAnchor );
+            break;
+          case QgsCallout::Centroid:
+          default:
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp]: Remove default

File: src/core/callouts/qgscallout.cpp [hunk 4]
```diff
@@ -412,7 +480,22 @@ void QgsManhattanLineCallout::draw( QgsRenderContext &context, QRectF rect, cons
...
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp]: Remove

File: src/core/callouts/qgscallout.h [hunk 5]
```diff
@@ -81,6 +82,15 @@ class CORE_EXPORT QgsCallout
       OrderBelowIndividualLabels, //!< Render callouts below their individual associated labels, some callouts may be drawn over other labels
     };
 
+    //! Feature's anchor point position
+    enum AnchorPoint
+    {
+      PoleOfInaccessibility = 0, //!< The surface's pole of inaccessibility used as anchor
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.h]: Might be worth marking these as polygon only options in the dox (planning ahead for future line anchor options)

File: src/core/callouts/qgscallout.h [hunk 6]
```diff
@@ -252,6 +262,37 @@ class CORE_EXPORT QgsCallout
...
+     * Returns the feature's anchor point position.
+     *
+     * \see setAnchorPoint()
+     */
+    AnchorPoint anchorPoint() const { return mAnchorPoint; }
+
+    /**
+     * Sets the feature's \a anchor point position.
+     *
+     * \see drawCalloutToAllParts()
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.h]: anchorPoint()

File: src/gui/callouts/qgscalloutwidget.cpp [hunk 7]
```diff
@@ -160,6 +160,13 @@ QgsSimpleLineCalloutWidget::QgsSimpleLineCalloutWidget( QgsVectorLayer *vl, QWid
 
   connect( mDrawToAllPartsCheck, &QCheckBox::toggled, this, &QgsSimpleLineCalloutWidget::drawToAllPartsToggled );
 
+  // Anchor point options
+  mAnchorPointComboBox->addItem( tr( "Pole of inaccessibility" ), static_cast< int >( QgsCallout::PoleOfInaccessibility ) );
```
Comment (nyalldawson) [inline src/gui/callouts/qgscalloutwidget.cpp]: Would be great to have icons for these in future

File: src/gui/callouts/qgscalloutwidget.cpp [hunk 8]
```diff
@@ -194,10 +201,19 @@ void QgsSimpleLineCalloutWidget::setCallout( QgsCallout *callout )
...
   registerDataDefinedButton( mMinCalloutLengthDDBtn, QgsCallout::MinimumCalloutLength );
   registerDataDefinedButton( mOffsetFromAnchorDDBtn, QgsCallout::OffsetFromAnchor );
   registerDataDefinedButton( mOffsetFromLabelDDBtn, QgsCallout::OffsetFromLabel );
   registerDataDefinedButton( mDrawToAllPartsDDBtn, QgsCallout::DrawCalloutToAllParts );
+  registerDataDefinedButton( mAnchorPointDDBtn, QgsCallout::AnchorPointPosition );
+}
+
+void QgsSimpleLineCalloutWidget::setGeometryType( QgsWkbTypes::GeometryType type )
+{
+  mAnchorPointComboBox->setEnabled( type == QgsWkbTypes::PolygonGeometry );
```
Comment (nyalldawson) [inline src/gui/callouts/qgscalloutwidget.cpp]: Maybe best to hide these too for non polygon layers?

File: src/gui/callouts/qgscalloutwidget.h [hunk 9]
```diff
@@ -81,6 +81,8 @@ class GUI_EXPORT QgsCalloutWidget : public QWidget, protected QgsExpressionConte
      */
     const QgsVectorLayer *vectorLayer() const { return mVectorLayer; }
 
+    virtual void setGeometryType( QgsWkbTypes::GeometryType type ) = 0;
```
Comment (nyalldawson) [inline src/gui/callouts/qgscalloutwidget.h]: Dox

**Review by nyalldawson — COMMENTED** [review]
Very nice!!

**Comment by nirvn** [comment]
@nyalldawson , thanks for the review, comments addressed and tests added.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**